File: src/main/java/toker/panel/aop/interceptor/DefaultItemValueInterceptor.java
package toker.panel.aop.interceptor;

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import toker.panel.entity.Item;
import toker.panel.repository.ItemRepository;
import toker.panel.annotation.DefaultItemValue;

import javax.inject.Inject;
import java.lang.reflect.Method;

@Aspect
public class DefaultItemValueInterceptor {

    private ItemRepository itemRepo;

    @Inject
    public void setItemRepo(ItemRepository itemRepo) {
        this.itemRepo = itemRepo;
    }

    @Pointcut("execution(public toker.panel.entity.Item *(..)) && " +
            "@annotation(toker.panel.annotation.DefaultItemValue)")
    public void getting() { }

    @Around("getting()")
    public Item aroundGetting(ProceedingJoinPoint joinPoint) throws Throwable {
        Item value = (Item) joinPoint.proceed();
        if (value == null) {
            Class<?> type = joinPoint.getSignature().getDeclaringType();
            try {
                Method getter = type.getDeclaredMethod(joinPoint.getSignature().getName());
                return itemRepo.getOne(getter.getAnnotation(DefaultItemValue.class).itemId());
            } catch (NoSuchMethodException ignored) { }
        }
        return value;
    }
}

File: src/main/java/toker/panel/aop/interceptor/DefaultItemValueInterceptor.kt
package toker.panel.aop.interceptor

import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import toker.panel.annotation.DefaultItemValue
import toker.panel.entity.Item
import toker.panel.repository.ItemRepository
import javax.inject.Inject

@Aspect
class DefaultItemValueInterceptor {

    lateinit var itemRepo: ItemRepository

    @Pointcut("execution(public toker.panel.entity.Item *(..)) && " +
            "@annotation(toker.panel.annotation.DefaultItemValue)")
    fun getting() { }

    @Around("getting()")
    fun aroundGetting(joinPoint: ProceedingJoinPoint): Item? {
        val value = joinPoint.proceed() as Item?
        if (value == null) {
            val type = joinPoint.signature.declaringType
            try {
                val getter = type.getDeclaredMethod(joinPoint.signature.name)
                return itemRepo!!.getOne(getter.getAnnotation(DefaultItemValue::class.java).itemId)
            } catch (ignored: NoSuchMethodException) { }
        }
        return value
    }
}